src/scene/layer-composition.js

```diff
@@ -466,31 +317,139 @@ class LayerComposition extends EventHandler {
             }
 
             this._renderActions.length = renderActionCount;
+        }
 
+        if ((result & COMPUPDATED_LIGHTS) || (result & COMPUPDATED_CAMERAS)) {
             this._logRenderActions();
         }
 
-        var arr;
-        if ((result & COMPUPDATED_LIGHTS) || (result & COMPUPDATED_CAMERAS)) {
-            // cameras/lights changed
-            this._globalLightCameraIds.length = 0;
-            for (l = 0; l < this._globalLightCameras.length; l++) {
-                arr = [];
-                for (i = 0; i < this._globalLightCameras[l].length; i++) {
-                    index = this.cameras.indexOf( this._globalLightCameras[l][i] );
-                    if (index < 0) {
-                        // #ifdef DEBUG
-                        console.warn("Can't find _globalLightCameras[l][i] in cameras");
-                        // #endif
-                        continue;
+        return result;
+    }
+
+    updateShadowCasters() {
+
+        // adjust _lightShadowCasters to the right size, matching number of lights, and clean it up
+        const lightCount = this._lights.length;
+        this._lightShadowCasters.length = lightCount;
+        for (let i = 0; i < lightCount; i++) {
+
+            let casters = this._lightShadowCasters[i];
+            if (casters) {
+                casters.clear();
+            } else {
+                this._lightShadowCasters[i] = new ShadowCasters();
+            }
+        }
+
+        // for each layer
+        const len = this.layerList.length;
+        for (let i = 0; i < len; i++) {
+            const layer = this.layerList[i];
+            const lights = layer._lights;
+
+            // for each light of a layer
+            for (let j = 0; j < lights.length; j++) {
+
+                // only need casters when casting shadows
+                if (lights[j].castShadows) {
+
+                    // find its index in global light list, and get shadow casters for it
+                    const lightIndex = this._lightsMap.get(lights[j]);
+                    const casters = this._lightShadowCasters[lightIndex];
+                    const castersSet = casters.set;
+                    const castersList = casters.list;
+
+                    // add unique meshes from the layer to casters
+                    const meshInstances = layer.shadowCasters;
+                    for (let k = 0; k < meshInstances.length; k++) {
+                        const meshInstance = meshInstances[k];
+
+                        if (!castersSet.has(meshInstance)) {
+                            castersSet.add(meshInstance);
+                            castersList.push(meshInstance);
```

Comment:
If you've taken the trouble to define a class to abstract/hide the Set+Array nature of the data structure, should you move `add` etc into `ShadowCasters` too?

Reply:
good suggestion and done!
I'm planning on expanding that class later when I implement shadow atlas to avoid re-rendering all shadow maps when casters don't change .. so this is a good start.